src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -121,99 +346,173 @@ private string CreateVaryByCacheKey(CachedVaryBy varyBy)
-
                 var cachedResponse = cacheEntry as CachedResponse;
+                var cachedResponseHeaders = new ResponseHeaders(cachedResponse.Headers);
 
-                var response = HttpContext.Response;
-                // Copy the cached status code and response headers
-                response.StatusCode = cachedResponse.StatusCode;
-                foreach (var pair in cachedResponse.Headers)
+                _responseTime = Clock.UtcNow;
+                var age = _responseTime - cachedResponse.Created;
+                age = age > TimeSpan.Zero ? age : TimeSpan.Zero;
+
+                if (EntryIsFresh(cachedResponseHeaders, age, verifyAgainstRequest: true))
                 {
-                    response.Headers[pair.Key] = pair.Value;
-                }
+                    var response = HttpContext.Response;
+                    // Copy the cached status code and response headers
+                    response.StatusCode = cachedResponse.StatusCode;
+                    foreach (var header in cachedResponse.Headers)
+                    {
+                        response.Headers.Add(header);
+                    }
 
-                // TODO: Update cache headers (Age)
-                response.Headers["Served_From_Cache"] = DateTime.Now.ToString();
+                    response.Headers[HeaderNames.Age] = age.TotalSeconds.ToString("F0", CultureInfo.InvariantCulture);
 
-                // Copy the cached response body
-                var body = cachedResponse.Body;
-                if (body.Length > 0)
+                    if (_responseType == ResponseType.HeadersOnly)
+                    {
+                        responseServed = true;
+                    }
+                    else if (_responseType == ResponseType.FullReponse)
+                    {
+                        // Copy the cached response body
+                        var body = cachedResponse.Body;
+
+                        // Add a content-length if required
+                        if (response.ContentLength == null && string.IsNullOrEmpty(response.Headers[HeaderNames.TransferEncoding]))
+                        {
+                            response.ContentLength = body.Length;
+                        }
+
+                        if (body.Length > 0)
+                        {
+                            await response.Body.WriteAsync(body, 0, body.Length);
+                        }
+
+                        responseServed = true;
+                    }
+                }
+                else
                 {
-                    await response.Body.WriteAsync(body, 0, body.Length);
+                    // TODO: Validate with endpoint instead
                 }
-                return true;
             }
 
-            return false;
-        }
+            if (!responseServed && RequestCacheControl.OnlyIfCached)
+            {
+                HttpContext.Response.Clear();
```

Comment:
Why do you need the Clear()? It looks like you only set headers in responseServed scenarios.